### ISSUE TITLE
Gometalinter improvements or terrible ideas

### ISSUE THREAD

**Issue by cep21** [alecthomas/gometalinter#31]
Hi,

I wrote https://github.com/cep21/goverify as part of my company's build process and would like to just replace it with gometalinter and think I could with a few improvements.  It is more macro based (https://github.com/cep21/goverify/blob/master/macros.go) with a goverify.json file configuring runs (https://github.com/signalfx/metricproxy/blob/master/goverify.json) where people run ```goverify -fix``` before a commit and the build script itself runs ```goverify``` and I can simply check the return code for failure.  Do you think the following improvements would belong inside gometalinter or do they not fully fit with gometalinter and I should keep goverify?

1. Add support for gofmt, goimports, and minimum code coverage .
2. Add a -fix flag for auto correcting format and imports
3. Add support for a gometalinter.json file that, if exists, specifies default arguments for gometalinter
4. Add support for ignoring directories or files when doing a ./... recursive directory argument
5. Add support for ignoring error messages/error code for messages that match a set of regexes.  For example, allow me to ignore golint errors that match regex ["GetID"].

**Comment by alecthomas** [comment]
Hi there,

> It is more macro based (https://github.com/cep21/goverify/blob/master/macros.go) with a goverify.json file configuring runs (https://github.com/signalfx/metricproxy/blob/master/goverify.json) where people run goverify -fix before a commit and the build script itself runs goverify and I can simply check the return code for failure.

gometalinter basically supports some (most?) of this via the `--linter=NAME:COMMAND:PATTERN` and `@<flags-file>` arguments. The former lets you define or replace linters and their arguments while the latter loads flags from files. So you could have a `gometalinter.flags` file containing a set of `--linter` definitions that will be run.

> 1. Add support for gofmt, goimports, and minimum code coverage .
> 2. Add a -fix flag for auto correcting format and imports

I think formatting/fixing code is outside the scope of what I want gometalinter to achieve, but minimum code coverage would definitely be nice to have. I don't think any of the existing coverage tooling will let you do this though, so someone will have to write it.

> 4. Add support for ignoring directories or files when doing a ./... recursive directory argument

This would *definitely* be useful to have.

> 5. Add support for ignoring error messages/error code for messages that match a set of regexes. For example, allow me to ignore golint errors that match regex ["GetID"].

This already exists:

```
  -e, --exclude=REGEXP
                       Exclude messages matching these regular expressions.
```

So in summary, I'd welcome some of the changes you suggest, but not support for formatting/fixing code as I think that's out of scope.

**Comment by tsenart** [comment]
What about verifying if code is formatted, instead of formatting it?

**Comment by alecthomas** [comment]
Yeah I don't have a problem with that at all.

**Comment by cep21** [comment]
For gofmt, the easiest solution is to run ```gofmt -s -l <FILENAME>``` and then print the filename as an error (maybe line zero to keep formatting consistent?) rather than try to figure out the exact line which is possible but not as trivial.  The same for ```goimports```.  But if you're going to add both of those, it's somewhat trivial to have a default off ```--fix``` flag that some users may appreciate.  

What do you mean by flags-file argument?  I don't see it in the readme.  What I enjoy about my ```goverify``` script is that it has simplified usage for developers of the code (a single command) once setup.

For coverage, it's also important to create a coverage profile since it's not always consistent and users need to see which lines missed coverage.

**Comment by alecthomas** [comment]
> For gofmt, the easiest solution is to run gofmt -s -l <FILENAME> and then print the filename as an error (maybe line zero to keep formatting consistent?) rather than try to figure out the exact line which is possible but not as trivial. The same for goimports.

Seems reasonable.

> What do you mean by flags-file argument? I don't see it in the readme. What I enjoy about my goverify script is that it has simplified usage for developers of the code (a single command) once setup.

If you have a file called `gometalinter.flags` with these contents:

```
--linter=vet:go tool vet -printfuncs=Infof,Debugf,Warningf,Errorf,Criticalf {path}/*.go:PATH:LINE:MESSAGE
--disable=dupl
--tests
--cyclo-over=15
```

Then if you run `gometalinter @gometalinter.flags` it will load all the command-line flags from that file.

> For coverage, it's also important to create a coverage profile since it's not always consistent and users need to see which lines missed coverage.

I did a bit of playing around yesterday and this bit of shell results in something that gometalinter could digest:

```
(go test -coverprofile=t.cov 2> /dev/null; go tool cover -func=t.cov) | tr -d '%' | awk '/\.go:/ && $NF < 60 {print $1, $NF}'
```

**Comment by cep21** [comment]
Thanks.  That makes sense.

1. I want to preserve simplicity of running.  Adding arguments increases the complexity over just ideally being able to run a command without any arguments (or few arguments) and having it usually do the right thing.  Is it possible to have gometalinter auto load a gometalinter.flags file if it exists in the running directory?
2. If run inside git, is it possible to autoload any gometalinter.flags files in parent directories of the git repository?
3. Is it possible to work on windows: not depend upon awk or other external things being installed other than go? (I saw awk in the above "go test" command)

**Comment by cep21** [comment]
I could implement 1 and 2, but want to know if they would be accepted

**Comment by alecthomas** [comment]
I think auto-loading flags from the "typical" locations `./.gometalinterrc` and `~/.gometalinterrc` would be a reasonable addition. I'm not sure that trying parent directories makes much sense in the broader context though?

I wrote [this](https://github.com/alecthomas/gocoverlint) a couple of days ago, but haven't quite finished adding support for it to gometalinter.

**Comment by cep21** [comment]
A current issue other developers are having, and myself, is that our codebase is very large.  Running checks on the entire thing is a bit slow.  It would be nice if I could be inside a sub directory and run checks on that part of the tree.  Then the problem becomes which checks to run.  By checking for a default configuration file in parent directories you can know which checks to run.  The second question becomes how far up parent directories do you check.  A reasonable suggestion there is however far the source code tree goes.

**Comment by alecthomas** [comment]
These seem like problems with solutions that probably don't fit all use-cases. A wrapper script seems like a better solution than shoe-horning logic into gometalinter. For example, if you use git submodules do you check them? Do you use parent configs in that case? What about support for non-git repositories?

If you only want to run gometalinter on packages with changed files you'd need some tool to find those packages from the changeset anyway. So adding any extra logic there, about what to run, seems like a better approach.

**Comment by cep21** [comment]
> if you use git submodules do you check them

It's simple enough to stop at the current repository's root.  Supporting non-git repositories could be a reasonable future improvement.  The goal is to make it easy to relint a directory while developing inside it.  However, wrapping gometalinter instead seems like a reasonable alternative.

```gocoverlint``` could use a timeout parameter it passes to tests: Maybe a good default would be gometalinter's timeout.  It shows coverage per function, which makes sense when thinking about a vim-ish parseable output, but I imagine most real world use cases would want coverage of the package as a whole to be over some percentage.

**Comment by alecthomas** [comment]
Having a flag to switch between package and function-level coverage would be good, agreed.

**Comment by alecthomas** [comment]
> However, wrapping gometalinter instead seems like a reasonable alternative.

This tool could be called `gometametalinter`!

**Comment by cep21** [comment]
Yo dawg I heard you like to lint code.

**Comment by cep21** [comment]
The only part missing with test coverage support is ignoring directories.  I imagine ```expandPaths()``` would be extended to take a []string that is every directory name to ignore.  An important one that comes to mind is ```vendor```.  Others in my case would be auto generated stuff around thrift or protocol buffers.

**Comment by alecthomas** [comment]
Yes agreed. A glob might be a good to support there.

**Comment by alecthomas** [comment]
I'm going to close this. Feel free to send PRs for issues we've discussed.
